File: benchmark/cpp/benchmark.cpp
// Copyright 2025 Robert Sayre
// Licensed under the Apache License, Version 2.0
// http://www.apache.org/licenses/LICENSE-2.0

#include <chrono>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

#include "rust/cpp-bindings/twitter.h"
#include "yaml-cpp/yaml.h"

const int ITERATIONS = 1000;
const int WARMUP_ITERATIONS = 100;

struct TestCase {
    std::string description;
    std::string text;
};

std::vector<TestCase> loadTests(const YAML::Node& node) {
    std::vector<TestCase> tests;
    for (const auto& test : node) {
        TestCase tc;
        if (test["description"]) {
            tc.description = test["description"].as<std::string>();
        }
        tc.text = test["text"].as<std::string>();
        tests.push_back(tc);
    }
    return tests;
}

double benchmarkAutolink(const std::vector<TestCase>& tests) {
    twitter_text::Autolinker autolinker;
    autolinker.setNoFollow(false);

    // Warmup
    for (int i = 0; i < WARMUP_ITERATIONS; i++) {
        for (const auto& test : tests) {
            autolinker.autolink(test.text);
        }
    }

    // Benchmark
    auto start = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < ITERATIONS; i++) {
        for (const auto& test : tests) {
            autolinker.autolink(test.text);
        }
    }
    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end - start;
    return ITERATIONS / elapsed.count();
}

double benchmarkExtract(
    const std::vector<TestCase>& mentions,
    const std::vector<TestCase>& urls,
    const std::vector<TestCase>& hashtags,
    const std::vector<TestCase>& cashtags
) {
    twitter_text::Extractor<> extractor;

    // Collect all texts (like JS/Rust benchmark)
    std::vector<std::string> allTexts;
    for (const auto& test : mentions) {
        allTexts.push_back(test.text);
    }
    for (const auto& test : urls) {
        allTexts.push_back(test.text);
    }
    for (const auto& test : hashtags) {
        allTexts.push_back(test.text);
    }
    for (const auto& test : cashtags) {
        allTexts.push_back(test.text);
    }

    // Warmup - call all 4 extract functions for each text (like JS/Rust)
    for (int i = 0; i < WARMUP_ITERATIONS; i++) {
        for (const auto& text : allTexts) {
            extractor.extractMentionedScreennamesWithIndices(text);
            extractor.extractUrlsWithIndices(text);
            extractor.extractHashtagsWithIndices(text);
            extractor.extractCashtagsWithIndices(text);
        }
    }

    // Benchmark
    auto start = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < ITERATIONS; i++) {
        for (const auto& text : allTexts) {
            extractor.extractMentionedScreennamesWithIndices(text);
            extractor.extractUrlsWithIndices(text);
            extractor.extractHashtagsWithIndices(text);
            extractor.extractCashtagsWithIndices(text);
        }
    }
    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end - start;
    return ITERATIONS / elapsed.count();
}

double benchmarkValidateTweet(const std::vector<TestCase>& tweets) {
    twitter_text::Validator validator;

    // Warmup
    for (int i = 0; i < WARMUP_ITERATIONS; i++) {
        for (const auto& test : tweets) {
            validator.isValidTweet(test.text);
        }
    }

    // Benchmark
    auto start = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < ITERATIONS; i++) {
        for (const auto& test : tweets) {
            validator.isValidTweet(test.text);
        }
    }
    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end - start;
    return ITERATIONS / elapsed.count();
}

double benchmarkValidateAll(
    const std::vector<TestCase>& tweets,
    const std::vector<TestCase>& usernames,
    const std::vector<TestCase>& hashtags,
    const std::vector<TestCase>& urls
) {
    twitter_text::Validator validator;

    // Warmup - call all 4 validate functions
    for (int i = 0; i < WARMUP_ITERATIONS; i++) {
        for (const auto& test : tweets) {
            validator.isValidTweet(test.text);
        }
        for (const auto& test : usernames) {
            validator.isValidUsername(test.text);
        }
        for (const auto& test : hashtags) {
            validator.isValidHashtag(test.text);
        }
        for (const auto& test : urls) {
            validator.isValidUrl(test.text);
        }
    }

    // Benchmark
    auto start = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < ITERATIONS; i++) {
        for (const auto& test : tweets) {
            validator.isValidTweet(test.text);
        }
        for (const auto& test : usernames) {
            validator.isValidUsername(test.text);
        }
        for (const auto& test : hashtags) {
            validator.isValidHashtag(test.text);
        }
        for (const auto& test : urls) {
            validator.isValidUrl(test.text);
        }
    }
    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end - start;
    return ITERATIONS / elapsed.count();
}

double benchmarkParse(const std::vector<TestCase>& tests) {
    twitter_text::TwitterTextConfiguration config;

    // Warmup
    for (int i = 0; i < WARMUP_ITERATIONS; i++) {
        for (const auto& test : tests) {
            twitter_text::TwitterTextParser::parse(test.text, config, true);
        }
    }

    // Benchmark
    auto start = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < ITERATIONS; i++) {
        for (const auto& test : tests) {
            twitter_text::TwitterTextParser::parse(test.text, config, true);
        }
    }
    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end - start;
    return ITERATIONS / elapsed.count();
}

int main(int argc, char* argv[]) {
    if (argc < 5) {
        std::cerr << "Usage: " << argv[0] 
                  << " <autolink.yml> <extract.yml> <validate.yml> <parse.yml>" 
                  << std::endl;
        return 1;
    }

    std::cout << "Twitter Text Benchmark: C++ (via Rust FFI)" << std::endl;
    std::cout << "===========================================" << std::endl;

    // Load YAML files
    YAML::Node autolinkYaml = YAML::LoadFile(argv[1]);
    YAML::Node extractYaml = YAML::LoadFile(argv[2]);
    YAML::Node validateYaml = YAML::LoadFile(argv[3]);
    YAML::Node parseYaml = YAML::LoadFile(argv[4]);

    // Load test cases
    auto autolinkTests = loadTests(autolinkYaml["tests"]);
    
    std::vector<TestCase> mentionTests, urlTests, hashtagTests, cashtagTests;
    if (extractYaml["tests"]["mentions"]) {
        mentionTests = loadTests(extractYaml["tests"]["mentions"]);
    }
    if (extractYaml["tests"]["urls"]) {
        urlTests = loadTests(extractYaml["tests"]["urls"]);
    }
    if (extractYaml["tests"]["hashtags"]) {
        hashtagTests = loadTests(extractYaml["tests"]["hashtags"]);
    }
    if (extractYaml["tests"]["cashtags"]) {
        cashtagTests = loadTests(extractYaml["tests"]["cashtags"]);
    }

    std::vector<TestCase> validateTweets, validateUsernames, validateHashtags, validateUrls;
    if (validateYaml["tests"]["tweets"]) {
        validateTweets = loadTests(validateYaml["tests"]["tweets"]);
    }
    if (validateYaml["tests"]["usernames"]) {
        validateUsernames = loadTests(validateYaml["tests"]["usernames"]);
    }
    if (validateYaml["tests"]["hashtags"]) {
        validateHashtags = loadTests(validateYaml["tests"]["hashtags"]);
    }
    if (validateYaml["tests"]["urls"]) {
        validateUrls = loadTests(validateYaml["tests"]["urls"]);
    }

    auto parseTests = loadTests(parseYaml["tests"]);

    // Run benchmarks
    double autolinkOps = benchmarkAutolink(autolinkTests);
    double extractOps = benchmarkExtract(mentionTests, urlTests, hashtagTests, cashtagTests);
    double validateTweetOps = benchmarkValidateTweet(validateTweets);
    double validateAllOps = benchmarkValidateAll(validateTweets, validateUsernames, validateHashtags, validateUrls);
    double parseOps = benchmarkParse(parseTests);

    // Print results
    std::cout << std::endl;
    std::cout << "Autolink (" << ITERATIONS << " iterations):" << std::endl;
    std::cout << "  C++: " << static_cast<int>(autolinkOps) << " ops/sec" << std::endl;

    std::cout << std::endl;
    std::cout << "Extract (" << ITERATIONS << " iterations):" << std::endl;
    std::cout << "  C++: " << static_cast<int>(extractOps) << " ops/sec" << std::endl;

    std::cout << std::endl;
    std::cout << "Validate Tweet (" << ITERATIONS << " iterations):" << std::endl;
    std::cout << "  C++: " << static_cast<int>(validateTweetOps) << " ops/sec" << std::endl;

    std::cout << std::endl;
    std::cout << "Validate All (" << ITERATIONS << " iterations):" << std::endl;
    std::cout << "  C++: " << static_cast<int>(validateAllOps) << " ops/sec" << std::endl;

    std::cout << std::endl;
    std::cout << "Parse Tweet (" << ITERATIONS << " iterations):" << std::endl;
    std::cout << "  C++: " << static_cast<int>(parseOps) << " ops/sec" << std::endl;

    std::cout << std::endl;
    std::cout << "Done." << std::endl;

    return 0;
}

File: benchmark/python/benchmark.py
#!/usr/bin/env python3
# Copyright 2025 Robert Sayre
# Licensed under the Apache License, Version 2.0
# http://www.apache.org/licenses/LICENSE-2.0

"""
Twitter Text Benchmark: Python (via Rust FFI)

Usage:
    benchmark.py [<autolink.yml> <extract.yml> <validate.yml> <parse.yml>]

If no arguments are provided, uses Bazel runfiles to find data files.
"""

import os
import sys
import time

import twitter_text
import yaml

ITERATIONS = 1000
WARMUP_ITERATIONS = 100


def get_runfiles_path(path: str) -> str:
    """Get the path to a file in Bazel runfiles."""
    # Try the runfiles directory relative to the script
    runfiles_dir = os.environ.get("RUNFILES_DIR")
    if runfiles_dir:
        full_path = os.path.join(runfiles_dir, "_main", path)
        if os.path.exists(full_path):
            return full_path

    # Try relative to script location
    script_dir = os.path.dirname(os.path.abspath(__file__))
    # Look for runfiles manifest
    for parent in [script_dir, os.path.dirname(script_dir)]:
        candidate = os.path.join(parent, "_main", path)
        if os.path.exists(candidate):
            return candidate
        # Also try without _main prefix
        candidate = os.path.join(parent, path)
        if os.path.exists(candidate):
            return candidate

    # Fallback: return the path as-is
    return path


def load_yaml(path: str) -> dict:
    with open(path, "r", encoding="utf-8") as f:
        return yaml.safe_load(f)


def benchmark_autolink(data: dict) -> float:
    autolinker = twitter_text.Autolinker()
    autolinker.set_no_follow(False)
    tests = data.get("tests", [])

    # Warmup
    for _ in range(WARMUP_ITERATIONS):
        for test in tests:
            autolinker.autolink(test["text"])

    # Benchmark
    start = time.perf_counter()
    for _ in range(ITERATIONS):
        for test in tests:
            autolinker.autolink(test["text"])
    elapsed = time.perf_counter() - start

    return ITERATIONS / elapsed


def benchmark_extract(data: dict) -> float:
    extractor = twitter_text.Extractor()
    tests = data.get("tests", {})

    # Collect all texts (like JS/Rust benchmark)
    all_texts = []
    for key in ["mentions", "urls", "hashtags", "cashtags"]:
        if key in tests and tests[key]:
            all_texts.extend(test["text"] for test in tests[key])

    # Warmup - call all 4 extract functions for each text
    for _ in range(WARMUP_ITERATIONS):
        for text in all_texts:
            extractor.extract_mentioned_screennames(text)
            extractor.extract_urls(text)
            extractor.extract_hashtags(text)
            extractor.extract_cashtags(text)

    # Benchmark
    start = time.perf_counter()
    for _ in range(ITERATIONS):
        for text in all_texts:
            extractor.extract_mentioned_screennames(text)
            extractor.extract_urls(text)
            extractor.extract_hashtags(text)
            extractor.extract_cashtags(text)
    elapsed = time.perf_counter() - start

    return ITERATIONS / elapsed


def benchmark_validate_tweet(data: dict) -> float:
    validator = twitter_text.Validator()
    tests = data.get("tests", {})
    tweets = tests.get("tweets", [])

    # Warmup
    for _ in range(WARMUP_ITERATIONS):
        for test in tweets:
            validator.is_valid_tweet(test["text"])

    # Benchmark
    start = time.perf_counter()
    for _ in range(ITERATIONS):
        for test in tweets:
            validator.is_valid_tweet(test["text"])
    elapsed = time.perf_counter() - start

    return ITERATIONS / elapsed


def benchmark_validate_all(data: dict) -> float:
    validator = twitter_text.Validator()
    tests = data.get("tests", {})
    tweets = tests.get("tweets", [])
    usernames = tests.get("usernames", [])
    hashtags = tests.get("hashtags", [])
    urls = tests.get("urls", [])

    # Warmup - call all 4 validate functions
    for _ in range(WARMUP_ITERATIONS):
        for test in tweets:
            validator.is_valid_tweet(test["text"])
        for test in usernames:
            validator.is_valid_username(test["text"])
        for test in hashtags:
            validator.is_valid_hashtag(test["text"])
        for test in urls:
            validator.is_valid_url(test["text"])

    # Benchmark
    start = time.perf_counter()
    for _ in range(ITERATIONS):
        for test in tweets:
            validator.is_valid_tweet(test["text"])
        for test in usernames:
            validator.is_valid_username(test["text"])
        for test in hashtags:
            validator.is_valid_hashtag(test["text"])
        for test in urls:
            validator.is_valid_url(test["text"])
    elapsed = time.perf_counter() - start

    return ITERATIONS / elapsed


def benchmark_parse(data: dict) -> float:
    config = twitter_text.TwitterTextConfiguration()
    tests = data.get("tests", [])

    # Warmup
    for _ in range(WARMUP_ITERATIONS):
        for test in tests:
            twitter_text.TwitterTextParser.parse(test["text"], config, True)

    # Benchmark
    start = time.perf_counter()
    for _ in range(ITERATIONS):
        for test in tests:
            twitter_text.TwitterTextParser.parse(test["text"], config, True)
    elapsed = time.perf_counter() - start

    return ITERATIONS / elapsed


def main():
    if len(sys.argv) >= 5:
        # Use command-line arguments
        autolink_path = sys.argv[1]
        extract_path = sys.argv[2]
        validate_path = sys.argv[3]
        parse_path = sys.argv[4]
    else:
        # Use Bazel runfiles
        autolink_path = get_runfiles_path("benchmark/data/autolink.yml")
        extract_path = get_runfiles_path("benchmark/data/extract.yml")
        validate_path = get_runfiles_path("benchmark/data/validate.yml")
        parse_path = get_runfiles_path("benchmark/data/parse.yml")

    print("Twitter Text Benchmark: Python (via Rust FFI)")
    print("==============================================")

    # Load YAML files
    autolink_data = load_yaml(autolink_path)
    extract_data = load_yaml(extract_path)
    validate_data = load_yaml(validate_path)
    parse_data = load_yaml(parse_path)

    # Run benchmarks
    autolink_ops = benchmark_autolink(autolink_data)
    extract_ops = benchmark_extract(extract_data)
    validate_tweet_ops = benchmark_validate_tweet(validate_data)
    validate_all_ops = benchmark_validate_all(validate_data)
    parse_ops = benchmark_parse(parse_data)

    # Print results
    print()
    print(f"Autolink ({ITERATIONS} iterations):")
    print(f"  Python: {int(autolink_ops)} ops/sec")

    print()
    print(f"Extract ({ITERATIONS} iterations):")
    print(f"  Python: {int(extract_ops)} ops/sec")

    print()
    print(f"Validate Tweet ({ITERATIONS} iterations):")
    print(f"  Python: {int(validate_tweet_ops)} ops/sec")

    print()
    print(f"Validate All ({ITERATIONS} iterations):")
    print(f"  Python: {int(validate_all_ops)} ops/sec")

    print()
    print(f"Parse Tweet ({ITERATIONS} iterations):")
    print(f"  Python: {int(parse_ops)} ops/sec")

    print()
    print("Done.")


if __name__ == "__main__":
    main()
